File: paleo_workbench/ui/pages/resource_table.py
"""工程资源清单表（Data 侧小表）。

V11 Model/View 迁移（goal §8 / 01-ui-audit D2）：此前 ``insertRow``-in-loop
逐格建 QTableWidgetItem 且页内重述表格 QSS（与 tokens.QSS_TEMPLATE 的
QTableView/QHeaderView 全局规则重复）。现改 ``QTableView + ObjectTableModel``
虚拟化，外观统一走全局 QSS + :func:`bind_table_defaults`；状态列着色经
``ColumnSpec.foreground_role`` 返回 token 名。
"""

from __future__ import annotations

from PySide6.QtCore import Qt
from PySide6.QtWidgets import QHeaderView, QTableView, QVBoxLayout, QWidget

from paleo_workbench.ui import tokens
from paleo_workbench.ui.components.states import PwbEmptyState
from paleo_workbench.ui.modelview import (
    ColumnSpec,
    ObjectTableModel,
    bind_table_defaults,
)

COLUMN_HEADERS = ["文件名", "类型", "格式", "状态", "路径"]
COLUMN_WIDTHS = [200, 100, 80, 100, 0]  # 0 = stretch


def _status_token(status: str) -> str:
    if status == "parsed":
        return "SUCCESS"
    if status == "error":
        return "ERROR_RED"
    return "TEXT_SECONDARY"


class _ResourceCellProxy:
    """Test-facing cell handle: ``item(row, col).text()`` without QTableWidgetItem."""

    def __init__(self, model, row: int, column: int) -> None:
        self._model = model
        self._row = row
        self._column = column

    def text(self) -> str:
        value = self._model.data(self._model.index(self._row, self._column))
        return "" if value is None else str(value)


class _ResourceTableView(QTableView):
    """QTableView + QTableWidget 兼容访问器（差分测试沿用 widget 式断言）。"""

    def rowCount(self) -> int:  # noqa: N802
        model = self.model()
        return 0 if model is None else model.rowCount()

    def columnCount(self) -> int:  # noqa: N802
        model = self.model()
        return 0 if model is None else model.columnCount()

    def item(self, row: int, column: int):
        model = self.model()
        if model is None:
            return None
        if row < 0 or column < 0 or row >= model.rowCount() or column >= model.columnCount():
            return None
        return _ResourceCellProxy(model, row, column)

    def horizontalHeaderItem(self, column: int):  # noqa: N802
        model = self.model()
        if model is None:
            return None

        class _Header:
            def __init__(self, text: str) -> None:
                self._text = text

            def text(self) -> str:
                return self._text

        return _Header(str(model.headerData(column, Qt.Orientation.Horizontal) or ""))


class ResourceTable(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("ResourceTable")
        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        self._model = ObjectTableModel(
            columns=[
                ColumnSpec("name", COLUMN_HEADERS[0], lambda res: str(getattr(res, "name", "") or "")),
                ColumnSpec(
                    "type",
                    COLUMN_HEADERS[1],
                    lambda res: tokens.RESOURCE_LABELS.get(
                        getattr(res, "type", ""), str(getattr(res, "type", ""))
                    ),
                ),
                ColumnSpec("format", COLUMN_HEADERS[2], lambda res: str(getattr(res, "format", "") or "")),
                ColumnSpec(
                    "status",
                    COLUMN_HEADERS[3],
                    lambda res: str(getattr(res, "status", "") or ""),
                    foreground_role=lambda res: _status_token(str(getattr(res, "status", "") or "")),
                ),
                ColumnSpec("path", COLUMN_HEADERS[4], lambda res: str(getattr(res, "path", "") or "")),
            ],
            key_of=lambda res: str(
                getattr(res, "id", "") or getattr(res, "path", "") or getattr(res, "name", "")
            ),
            parent=self,
        )
        self.table = _ResourceTableView()
        self.table.setModel(self._model)
        bind_table_defaults(self.table)
        header = self.table.horizontalHeader()
        for i, w in enumerate(COLUMN_WIDTHS):
            if w > 0:
                header.resizeSection(i, w)
            else:
                header.setSectionResizeMode(i, QHeaderView.ResizeMode.Stretch)
        layout.addWidget(self.table)

        # 空资源清单的统一空态（词汇与任务中心一致）。
        self._empty_state = PwbEmptyState("暂无资源", "导入数据后将在此列出工程资源。", parent=self)
        self._empty_state.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents)
        self._empty_state.setParent(self.table)
        self._empty_state.hide()
        self._model.modelReset.connect(self._update_empty_state)
        self._model.rowsInserted.connect(self._update_empty_state)
        self._model.rowsRemoved.connect(self._update_empty_state)

    def update_resources(self, resources: list) -> None:
        self._model.set_rows(list(resources or []))
        self._update_empty_state()

    def resizeEvent(self, event) -> None:  # noqa: N802
        super().resizeEvent(event)
        if self._empty_state.isVisible():
            self._empty_state.setGeometry(self.table.viewport().rect())

    def _update_empty_state(self, *_args) -> None:
        if self._model.rowCount() == 0:
            self._empty_state.setGeometry(self.table.viewport().rect())
            self._empty_state.show()
            self._empty_state.raise_()
        else:
            self._empty_state.hide()

File: libs/ui_pages_data/src/qt/resource_table.cpp
// UI-06 — ResourceTable shell (see qt/resource_table.hpp).
#include <pwb/ui_pages_data/qt/resource_table.hpp>

#include <QHeaderView>
#include <QLabel>
#include <QVBoxLayout>

#include <pwb/ui_pages_data/table_model.hpp>
#include <pwb/ui_pages_data/vocab.hpp>
#include <pwb/ui_shell/style_registry.hpp>

namespace pwb::ui_pages_data::qt {
namespace {

// COLUMN_HEADERS / COLUMN_WIDTHS (0 = stretch).
constexpr const char* kHeaders[] = {"文件名", "类型", "格式", "状态", "路径"};
constexpr int kWidths[] = {200, 100, 80, 100, 0};

QString pal_color(const char* token) {
    const auto p = ui_shell::style_palette();
    const auto it = p.find(token);
    return it != p.end() ? QString::fromStdString(it->second) : QString();
}

}  // namespace

ResourceTableModel::ResourceTableModel(QObject* parent)
    : QAbstractTableModel(parent) {}

void ResourceTableModel::set_rows(std::vector<AssetView> rows) {
    beginResetModel();
    rows_ = std::move(rows);
    endResetModel();
}

const AssetView* ResourceTableModel::row_at(int row) const {
    if (row < 0 || row >= static_cast<int>(rows_.size())) return nullptr;
    return &rows_[static_cast<std::size_t>(row)];
}

std::string ResourceTableModel::key_for_row(int row) const {
    const AssetView* view = row_at(row);
    if (view == nullptr) return {};
    if (!view->id.empty()) return view->id;
    if (!view->path.empty()) return view->path;
    return view->name;
}

int ResourceTableModel::index_for_key(const std::string& key) const {
    for (std::size_t i = 0; i < rows_.size(); ++i) {
        if (key_for_row(static_cast<int>(i)) == key)
            return static_cast<int>(i);
    }
    return -1;
}

int ResourceTableModel::rowCount(const QModelIndex& parent) const {
    return parent.isValid() ? 0 : static_cast<int>(rows_.size());
}

int ResourceTableModel::columnCount(const QModelIndex& parent) const {
    return parent.isValid() ? 0 : 5;
}

QVariant ResourceTableModel::data(const QModelIndex& index, int role) const {
    const AssetView* view = row_at(index.row());
    if (view == nullptr || index.column() < 0 || index.column() >= 5) {
        return {};
    }
    if (role == Qt::ItemDataRole::DisplayRole) {
        switch (index.column()) {
        case 0: return QString::fromStdString(view->name);
        case 1:
            return QString::fromStdString(
                std::string(resource_label(view->type)));
        case 2: return QString::fromStdString(view->format);
        case 3: return QString::fromStdString(view->status);
        case 4: return QString::fromStdString(view->path);
        default: return {};
        }
    }
    if (role == Qt::ItemDataRole::ForegroundRole && index.column() == 3) {
        const std::string token =
            std::string(status_color_token(view->status));
        const QString color = pal_color(token.c_str());
        if (!color.isEmpty()) return QColor(color);
    }
    if (role == Qt::ItemDataRole::UserRole && index.column() == 0) {
        return QString::fromStdString(key_for_row(index.row()));
    }
    return {};
}

QVariant ResourceTableModel::headerData(int section,
                                        Qt::Orientation orientation,
                                        int role) const {
    if (orientation == Qt::Orientation::Horizontal &&
        role == Qt::ItemDataRole::DisplayRole && section >= 0 && section < 5) {
        return QString::fromUtf8(kHeaders[section]);
    }
    return {};
}

// ---------------------------------------------------------------------------

ResourceTable::ResourceTable(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("ResourceTable"));
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    model_ = new ResourceTableModel(this);
    table_ = new QTableView(this);
    table_->setModel(model_);
    // bind_table_defaults verbatim.
    table_->setAlternatingRowColors(true);
    table_->setShowGrid(false);
    table_->setSelectionBehavior(QTableView::SelectionBehavior::SelectRows);
    table_->setEditTriggers(QTableView::EditTrigger::NoEditTriggers);
    table_->verticalHeader()->setVisible(false);
    table_->horizontalHeader()->setStretchLastSection(false);
    table_->horizontalHeader()->setHighlightSections(false);
    QHeaderView* header = table_->horizontalHeader();
    for (int i = 0; i < 5; ++i) {
        if (kWidths[i] > 0) {
            header->resizeSection(i, kWidths[i]);
        } else {
            header->setSectionResizeMode(
                i, QHeaderView::ResizeMode::Stretch);
        }
    }
    layout->addWidget(table_);

    // PwbEmptyState seam (ui.components.states) — same title/hint pair.
    empty_state_ = new QLabel(
        QStringLiteral("暂无资源\n导入数据后将在此列出工程资源。"), table_);
    empty_state_->setAlignment(Qt::AlignmentFlag::AlignCenter);
    empty_state_->setAttribute(Qt::WidgetAttribute::WA_TransparentForMouseEvents);
    empty_state_->hide();
    connect(model_, &QAbstractItemModel::modelReset, this,
            &ResourceTable::update_empty_state);
    connect(model_, &QAbstractItemModel::rowsInserted, this,
            &ResourceTable::update_empty_state);
    connect(model_, &QAbstractItemModel::rowsRemoved, this,
            &ResourceTable::update_empty_state);
}

void ResourceTable::update_resources(std::vector<AssetView> resources) {
    model_->set_rows(std::move(resources));
    update_empty_state();
}

void ResourceTable::resizeEvent(QResizeEvent* event) {
    QWidget::resizeEvent(event);
    if (empty_state_->isVisible()) {
        empty_state_->setGeometry(table_->viewport()->rect());
    }
}

void ResourceTable::update_empty_state() {
    if (model_->rowCount() == 0) {
        empty_state_->setGeometry(table_->viewport()->rect());
        empty_state_->show();
        empty_state_->raise();
    } else {
        empty_state_->hide();
    }
}

}  // namespace pwb::ui_pages_data::qt
